### PR TITLE
Added key expression support to Kafka producer

### LINKED ISSUE #134: Add support for setting the message key via SpEL
As a developer, I want to be able to configure a SpEL expression for automatically populating the outbound message key.

### PR DESCRIPTION
**Pull request by hekonsek**
This is an attempt to support dynamic Kafka keys expressions as described in - https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/issues/130 .

Fix #134

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@hekonsek Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-stream-binder-kafka&pullRequestId=131)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-stream-binder-kafka&pullRequestId=131) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by mbogoevici** [comment]
@hekonsek - thanks for the contribution. Looks good, in general. Would you mind to add a test too?

Thanks!

**Comment by hekonsek** [comment]
Sure :) Just did it.

**Comment by pivotal-issuemaster** [comment]
@hekonsek Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-stream-binder-kafka&pullRequestId=131)!

**Comment by hekonsek** [comment]
Please let me know if you want me to submit a documentation for this option to spring-cloud-stream AsciiDoc.

**Comment by mbogoevici** [comment]
@hekonsek Yes, please add some documentation. We maintain it in the same repository as the binder, i.e. https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/blob/master/spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc so it can just be an extra commit to this PR.

**Comment by hekonsek** [comment]
Done!

File: spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc [hunk 1]
```diff
@@ -190,6 +190,10 @@ batchTimeout::
 (Normally the producer does not wait at all, and simply sends all the messages that accumulated while the previous send was in progress.) A non-zero value may increase throughput at the expense of latency.
 +
 Default: `0`.
+messageKeyExpression::
+Expression (executed against incoming message) used to resolve the key of the produced Kafka message. For example `headers.key` or `payload.myKey`.
```
Comment (mbogoevici) [inline spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc]: Actually it is the outgoing message (from the application's perspective). We can indicate it is a 'SpEL' expression too (will fix on merge).

**Comment by mbogoevici** [comment]
Another quick comment is that we add author tags for significant changes and update the copyright years on all changes - will fix on merge.

**Comment by mbogoevici** [comment]
Squashed rebased and merged as https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/commit/61e7936978180e194d0ce67c97e4dc2d81cdbdfb.

Additional documentation polishing via https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/commit/c68ea8c570a719b2a1bcd9febcd2feab0d3dd8a2.

**Comment by mbogoevici** [comment]
Thanks for the contribution @hekonsek - looking forward for more :)

**Comment by hekonsek** [comment]
Many thanks for all your help Marius. 

I will have some more questions for you regarding taking advantage of this improvement and dynamic destinations in Data Flow environment. Which eventually may end up in additional pull requests as well. But I will open another issues for this problem and ping your from there :) .

**Pull request closed without merging**